src/w_wad.c

```diff
@@ -80,393 +70,118 @@ void ExtractFileBase(const char *path, char *dest)
 }
 
-static void AddWadInZip(mz_zip_archive *zip, const char *name, int index,
-                        size_t data_size)
+boolean W_SkipFile(const char *filename)
 {
-    byte *data = malloc(data_size);
-
-    if (!mz_zip_reader_extract_to_mem(zip, index, data, data_size, 0))
-    {
-        I_Error("AddWadInZip: mz_zip_reader_extract_to_mem failed");
-    }
-
-    I_Printf(VB_INFO, " adding %s", name);
-
-    wadinfo_t header;
-
-    if (sizeof(header) > data_size)
-    {
-        I_Error("Error reading header from %s", name);
-    }
-
-    memcpy(&header, data, sizeof(header));
-
-    if (strncmp(header.identification, "IWAD", 4)
-        && strncmp(header.identification, "PWAD", 4))
-    {
-        I_Error("Wad file %s doesn't have IWAD or PWAD id", name);
-    }
-
-    header.numlumps = LONG(header.numlumps);
-    if (header.numlumps == 0)
-    {
-        I_Printf(VB_WARNING, "Wad file %s is empty", name);
-        free(data);
-        return;
-    }
-
-    header.infotableofs = LONG(header.infotableofs);
-    if (header.infotableofs + header.numlumps * sizeof(filelump_t) > data_size)
-    {
-        I_Printf(VB_WARNING, "Error seeking offset from %s", name);
-        free(data);
-        return;
-    }
-
-    filelump_t *fileinfo = (filelump_t *)(data + header.infotableofs);
+    static const char *ext[] = { ".wad", ".zip", ".pk3", ".deh", ".txt" };
 
-    const char *wadname = M_StringDuplicate(name);
-    array_push(wadfiles, wadname);
-
-    numlumps += header.numlumps;
-
-    array_grow(lumpinfo, numlumps);
-
-    for (int i = 0; i < header.numlumps; i++)
+    for (int i = 0; i < arrlen(ext); ++i)
     {
-        lumpinfo_t item = {0};
-        M_CopyLumpName(item.name, fileinfo[i].name);
-        int size = LONG(fileinfo[i].size);
-        int position = LONG(fileinfo[i].filepos);
-        if (position + size > data_size)
+        if (M_StringCaseEndsWith(filename, ext[i]))
         {
-            I_Error("Error reading lump %d from %s", i, wadname);
+            return true;
         }
-        item.size = size;
-        item.data = data + position;
-        item.namespace = ns_global;
- 
-        // [FG] WAD file that contains the lump
-        item.wad_file = wadname;
-        array_push(lumpinfo, item);
     }
+    return false;
 }
 
-static void AddMarker(const char *name)
+static struct
 {
-    lumpinfo_t marker = {0};
-    M_CopyLumpName(marker.name, name);
-    array_push(lumpinfo, marker);
-    numlumps++;
-}
+    const char *dir;
+    const char *start_marker;
+    const char *end_marker;
+    namespace_t namespace;
+} subdirs[] = {
+    {"music",     NULL,       NULL,     ns_global   },
+    {"sprites",   "S_START",  "S_END",  ns_sprites  },
+    {"flats",     "F_START",  "F_END",  ns_flats    },
+    {"colormaps", "C_START",  "C_END",  ns_colormaps},
+    {"voxels",    "VX_START", "VX_END", ns_voxels   },
+};
+
+static struct
+{
+    const char *dir;
+    GameMode_t mode;
+    GameMission_t mission;
+} filters[] = {
```

Comment:
What exactly are these filters? Are they documented anywhere?

Reply:
Documentation: https://zdoom.org/wiki/Lump_filtering

These are needed for VoxelDoom_v_2.3.pk3 for example. We used to just load each file in pk3, but now there are possible conflicts with some PNG files.

Reply:
I see, thanks!